File: 0198-House-Robber/java-0198/src/Solution2.java
/// Source : https://leetcode.com/problems/house-robber/description/
/// Author : liuyubobobo
/// Time   : 2017-11-19

/// Dynamic Programming
/// Time Complexity: O(n)
/// Space Complexity: O(n)
public class Solution2 {

    public int rob(int[] nums) {

        int n = nums.length;
        if(n == 0)
            return 0;

        // the max profit for robbing nums[i...n)
        int[] memo = new int[nums.length];
        memo[n - 1] = nums[n - 1];
        for(int i = n - 2 ; i >= 0 ; i --)
            memo[i] = Math.max(memo[i + 1],
                               nums[i] + (i + 2 < n ? memo[i + 2] : 0));

        return memo[0];
    }

    public static void main(String[] args) {

        int nums[] = {2, 1};
        System.out.println((new Solution2()).rob(nums));
    }
}

File: 0001-0500/0152-Maximum-Product-Subarray/py-0152/Solution2.py
# Source : https://leetcode.com/problems/maximum-product-subarray/
# Author : penpenps
# Time   : 2019-07-21

from typing import List

# If even number of elements, the max product is multiple all nums together
# If odd number, it should be divided into to part from one of the negative numbers and the answer is one of two seperated parts to multiple together
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        rnums = nums[::-1]
        for i in range(1, n):
            nums[i] *= nums[i-1] or 1
            rnums[i] *= rnums[i-1] or 1

        return max(nums + rnums)
        

if __name__ == '__main__':
	s = Solution()
	nums = [2,3,-2,4]
	answer = s.maxProduct(nums)
	print(answer)
